Add MSAA example

rukai opened the pull request:
This example is the same as the triangle example except:
*    The render pipeline is created with a sample_count of 4
*    A new texture with a sample_count of 4 is created and set as the color_attachment instead of the swapchain
*    The swapchain is now specified as a resolve_target

It doesn't work yet as wgpu-native doesn't handle MSAA yet.

kvark commented:
This could be a good start! We need to make sure the examples don't repeat themselves too much and try to cover more ground. For this case, perhaps we could think of some workload involving a bunch of lines to draw? That would show MSAA as well as non-triangular primitive topologies.

kvark commented:
Also, let's use the framework module to avoid boilerplate repetition, like cube and shadow examples do. I don't think there needs to be more than hello-triangle and hello-compute for barebone setup.

rukai commented:
Ah right, I missed the framework.rs. I would appreciate it if I could get feedback as is, about if I am correctly using the wgpu API for MSAA. Then I'll go ahead and make the changes you have requested.

aloucks commented on examples/msaa/main.rs:
```diff
@@ -0,0 +1,181 @@
+        {
+            let mut rpass = encoder.begin_render_pass(&wgpu::RenderPassDescriptor {
+                color_attachments: &[wgpu::RenderPassColorAttachmentDescriptor {
+                    attachment: &multisampled_frame.create_default_view(),
```
Should the texture and view be created every frame or only when the window size changes?

kvark replied:
It is a good question, and it's currently unclear because Google is pushing for caching all the things by the implementation, i.e. the device would remember if there is already a good view for a given texture and return it again.
I'd prefer if we didn't rely on this hypothetical behavior :)

kvark commented:
> I would appreciate it if I could get feedback as is, about if I am correctly using the wgpu API for MSAA

I think it's correct. We might be missing some implementation bits (like https://github.com/gfx-rs/gfx/pull/2813), but it would be easy to fix them given this test case :)

rukai commented:
I changed the example to use PrimitiveTopology::LineList and to utilize the framework module.

I also added runtime configurable sample_count.

![Screenshot from 2019-06-20 00-06-32](https://user-images.githubusercontent.com/5120858/59773050-4e998100-92f0-11e9-8912-cc70f4d8abd7.png)

If you want to test the rendering then you will need to disable the MSAA logic by setting the swapchain as the color attachment rather then the resolve_target and you will need to set the sample_count to 1.

kvark commented:
@rukai this is a great test case, thank you!

kvark commented on examples/msaa-linelist/main.rs:
```diff
@@ -0,0 +1,235 @@
+            .create_buffer_mapped(vertex_data.len(), wgpu::BufferUsage::VERTEX)
+            .fill_from_slice(&vertex_data);
+        let index_buffer = device
+            .create_buffer_mapped(index_data.len(), wgpu::BufferUsage::VERTEX)
```
should be `INDEX`

kvark commented on examples/msaa-linelist/main.rs:
```diff
@@ -0,0 +1,235 @@
+                        Some(wgpu::winit::VirtualKeyCode::Left) => {
+                            if self.sample_count >= 2 {
+                                self.sample_count = self.sample_count >> 1;
+                                self.pipeline = Example::pipeline(device, &self.vs_module, &self.fs_module, &self.pipeline_layout, self.sample_count);
```
could we keep `update` function signature unchanged (i.e. not getting the `Device` parameter) by just storing the `requested_sample_count` variable and updating the actual `pipeline` at the start of `render`?

kvark commented on examples/msaa-linelist/main.rs:
```diff
@@ -0,0 +1,235 @@
+                            }
+                        }
+                        Some(wgpu::winit::VirtualKeyCode::Right) => {
+                            if self.sample_count <= 32 {
```
I don't think we want 64 samples supported (D3D12/Metal limitations...). Let's change this to `<32`

kvark commented on examples/msaa-linelist/main.rs:
```diff
@@ -0,0 +1,235 @@
+        {
+            let mut rpass = encoder.begin_render_pass(&wgpu::RenderPassDescriptor {
+                color_attachments: &[wgpu::RenderPassColorAttachmentDescriptor {
+                    attachment: &self.multisampled_framebuffer.create_default_view(),
```
let's keep this view cached in `self`

kvark commented on examples/msaa-linelist/main.rs:
```diff
@@ -0,0 +1,235 @@
+                    resolve_target: Some(&frame.view),
+                    load_op: wgpu::LoadOp::Clear,
+                    store_op: wgpu::StoreOp::Store,
+                    clear_color: wgpu::Color {
```
let's use `Color::BLACK`

kvark commented on examples/msaa-linelist/main.rs:
```diff
@@ -0,0 +1,235 @@
+            },
+            primitive_topology: wgpu::PrimitiveTopology::LineList,
+            color_states: &[wgpu::ColorStateDescriptor {
+                format: wgpu::TextureFormat::Bgra8Unorm,
```
let's use the `format` provided to use in `SwapChainDescriptor`

kvark commented on examples/msaa-linelist/main.rs:
```diff
@@ -0,0 +1,235 @@
+}
+
+impl Example {
+    fn pipeline(device: &mut wgpu::Device, vs_module: &wgpu::ShaderModule, fs_module: &wgpu::ShaderModule, pipeline_layout: &wgpu::PipelineLayout, sample_count: u32) -> wgpu::RenderPipeline {
```
let's rename this to start with a verb, i.e. `create_pipeline`

kvark replied:
here and in the other function we need to only borrow `&Device` immutably

kvark commented on examples/msaa-linelist/main.rs:
```diff
@@ -0,0 +1,235 @@
+                step_mode: wgpu::InputStepMode::Vertex,
+                attributes: &[
+                    wgpu::VertexAttributeDescriptor {
+                        format: wgpu::VertexFormat::Float4,
```
should be `Float2`

kvark commented on examples/msaa-linelist/main.rs:
```diff
@@ -0,0 +1,235 @@
+        })
+    }
+
+    fn multisampled_framebuffer(device: &mut wgpu::Device, width: u32, height: u32, sample_count: u32) -> wgpu::Texture {
```
similarly, let's rename this to start with a verb

kvark commented on examples/msaa-linelist/main.rs:
```diff
@@ -0,0 +1,235 @@
+            mip_level_count: 1,
+            sample_count: sample_count,
+            dimension: wgpu::TextureDimension::D2,
+            format: wgpu::TextureFormat::Bgra8Unorm,
```
same here - need to use format given to us in the swapchain descriptor

kvark commented on examples/msaa-linelist/main.rs:
```diff
@@ -0,0 +1,235 @@
+        let vs_module = device.create_shader_module(&vs_bytes);
+        let fs_module = device.create_shader_module(&fs_bytes);
+
+        let bind_group_layout =
```
let's not have any bindings or bind groups at all, since they are not needed! That is also a good property to show off in this example

kvark replied:
Let's leave it there. The logic here is: having no bind groups in a pipeline layout is an *edge* case. So it needs to be shown somewhere, but not necessarily in the hello-triangle example.

rukai replied:
Sounds good to me. I reverted the hello-triangle change.

kvark commented on examples/msaa-linelist/main.rs:
```diff
@@ -0,0 +1,235 @@
+                _color: [sin, -cos, 1.0, 1.0],
+            });
+
+            index_data.extend(&[
```
let's not have index buffer at all, since it's not saving us anything here

rukai commented on examples/msaa-linelist/shader.vert:
```diff
@@ -0,0 +1,10 @@
+#version 450
+
+layout(location = 0) in vec2 a_Pos;
+layout(location = 1) in vec4 a_color;
```
Are you just asking for the capitalization? I'm not sure what the first character is supposed to be.

kvark replied:
yes, capitalization of the letter after `_`. The first character here stands for "Attribute", fyi

rukai commented:
Thanks for the fantastic feedback!
I even learned something to apply to my own project.

I've made a few clarification questions and addressed everything else.
